### PR TITLE
[HOPSWORKS-2645] Support feature validation for on-demand FGs in hsfs

### PR DESCRIPTION
**Pull request by tkakantousis**

### CODE REVIEW

File: python/hsfs/core/feature_group_base_engine.py [hunk 1]
```diff
@@ -52,3 +52,14 @@ def update_statistics_config(self, feature_group):
         self._feature_group_api.update_metadata(
             feature_group, feature_group, "updateStatsConfig"
         )
+
+    def update_config(self, feature_group, metadata):
+        """Update the metadata attribute specified of the feature group ."""
+        metadata_values = {
+            "updateStatsSettings": True,
+            "validationType": feature_group.validation_type,
+        }
```
Comment (moritzmeister) [inline python/hsfs/core/feature_group_base_engine.py:61]: this looks like old code to me, I actually think we forgot to remove the function previously, cause look at line 50 above.

File: python/hsfs/core/feature_group_engine.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
 class FeatureGroupEngine(feature_group_base_engine.FeatureGroupBaseEngine):
     def save(self, feature_group, feature_dataframe, write_options):
-
+        print("FeatureGroupEngine.save()")
```
Comment (moritzmeister) [inline python/hsfs/core/feature_group_engine.py:26]: do we really want these prints?

File: python/hsfs/core/feature_group_engine.py [hunk 3]
```diff
@@ -154,17 +148,6 @@ def commit_delete(feature_group, delete_df, write_options):
         )
         return hudi_engine_instance.delete_record(delete_df, write_options)
 
-    def update_config(self, feature_group, metadata):
```
Comment (moritzmeister) [inline python/hsfs/core/feature_group_engine.py:157]: I think this function was old and not actually used anymore, so you shouldn't need it in the base class. but please double check.

**Review by moritzmeister — CHANGES_REQUESTED** [review]
You should be able to move the duplicated functions to the feature group base class.

File: python/hsfs/core/feature_group_base_engine.py (hunk 1 above)
Comment (tkakantousis) [reply python/hsfs/core/feature_group_base_engine.py:61]: It's still used here https://github.com/logicalclocks/feature-store-api/blob/ee70fecba586727d5ae9e5cc1d2e10463fad9819/python/hsfs/feature_group.py#L1224 

do you mean we should have an update_validation function?

File: java/src/main/java/com/logicalclocks/hsfs/OnDemandFeatureGroup.java [hunk 4]
```diff
@@ -83,6 +89,12 @@ public OnDemandFeatureGroup(FeatureStore featureStore, @NonNull String name, Int
     this.storageConnector = storageConnector;
     this.features = features;
     this.statisticsConfig = statisticsConfig != null ? statisticsConfig : new StatisticsConfig();
+    this.validationType = validationType != null ? validationType : ValidationType.NONE;
+    if (expectations != null && !expectations.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply java/src/main/java/com/logicalclocks/hsfs/OnDemandFeatureGroup.java:93]: there is not Builder for the base class, which constructor would I add this attributes to?

File: java/src/main/java/com/logicalclocks/hsfs/engine/OnDemandFeatureGroupEngine.java [hunk 5]
```diff
@@ -32,13 +33,18 @@
 
   public void saveFeatureGroup(OnDemandFeatureGroup onDemandFeatureGroup)
       throws FeatureStoreException, IOException {
+    Dataset<Row> onDemandDataset = null;
     if (onDemandFeatureGroup.getFeatures() == null) {
-      Dataset<Row> onDemandDataset = SparkEngine.getInstance()
+      onDemandDataset = SparkEngine.getInstance()
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply java/src/main/java/com/logicalclocks/hsfs/engine/OnDemandFeatureGroupEngine.java:38]: By "this" do you mean registering the temp table? If so, I don't know, that's out of scope of this PR. That is currently on master https://github.com/logicalclocks/feature-store-api/blob/ce659adcfbc54a957da9f6ad041e950809f42285/java/src/main/java/com/logicalclocks/hsfs/engine/OnDemandFeatureGroupEngine.java#L35

this PR just adds the validation at the end of the saveFeatureGroup method

File: python/hsfs/storage_connector.py [hunk 6]
```diff
@@ -705,7 +705,9 @@ def connection_string(self):
 
     @property
     def arguments(self):
-        """Additional JDBC, REDSHIFT, or Snowflake arguments."""
+        """Additional JDBC, REDSHIFT, or Snowflake arguments. When running hsfs with PySpark/Spark in Hopsworks,
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply python/hsfs/storage_connector.py]: fixed

File: python/hsfs/core/feature_group_engine.py (hunk 2 above)
Comment (tkakantousis) [reply python/hsfs/core/feature_group_engine.py]: fixed

File: python/hsfs/core/feature_group_base_engine.py (hunk 1 above)
Comment (tkakantousis) [reply python/hsfs/core/feature_group_base_engine.py:61]: I will try to use the update_metadata function

File: java/src/main/java/com/logicalclocks/hsfs/metadata/FeatureGroupBase.java [hunk 7]
```diff
@@ -80,8 +85,21 @@
   @Setter
   protected StatisticsConfig statisticsConfig = new StatisticsConfig();
 
+  @Getter
+  @Setter
+  @JsonProperty("validationType")
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply java/src/main/java/com/logicalclocks/hsfs/metadata/FeatureGroupBase.java]: fixed

File: python/hsfs/core/feature_group_base_engine.py (hunk 1 above)
Comment (tkakantousis) [reply python/hsfs/core/feature_group_base_engine.py]: fixed

File: python/hsfs/core/feature_group_engine.py (hunk 3 above)
Comment (tkakantousis) [reply python/hsfs/core/feature_group_engine.py:157]: fixed

File: java/src/main/java/com/logicalclocks/hsfs/OnDemandFeatureGroup.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (moritzmeister) [reply java/src/main/java/com/logicalclocks/hsfs/OnDemandFeatureGroup.java:93]: you are right

**Review by moritzmeister — APPROVED** [review]

**Pull request merged**